src/Refinery/DeriveTransformFromApplyTo.php

```diff
@@ -0,0 +1,37 @@
+<?php
+/* Copyright (c) 1998-2019 ILIAS open source, Extended GPL, see docs/LICENSE */
+
+/**
+ * @author  Niels Theen <ntheen@databay.de>
+ */
+
+namespace ILIAS\Refinery;
+
+
+use ILIAS\Data\Result;
+use ILIAS\DI\Exceptions\Exception;
+use ILIAS\Refinery\Validation\Constraints\ConstraintViolationException;
+
+trait DeriveTransformFromApplyTo
+{
+	/**
+	 * @param mixed $from
+	 * @return Result
+	 */
+	public function transform($from)
+	{
+		/** @var Result $result */
+		$result = $this->applyTo(new Result\Ok($from));
+		if (true === $result->isError()) {
+			$error = $result->error();
+
+			$message = $error;
+			if ($error instanceof Exception) {
+				$message = $error->getMessage();
```

Comment:
```suggestion
				throw $error;
```

This might be anything, we don't really know if this is a Constraint Violation. Same for the case, where error is a `string`, but I am currently not sure, which exception we would want to use for that case. I would *not* use ConstraintViolation, because this would be caught by some mechanism later and produce a user-facing mechanism that could only say "some error". I would rather let this bubble to the outer error-catching layer, which would produce a whoops or a message via error.php.

Reply:
@mjansenDatabay and I discussed this and we are generally with you. Bubbling this exception up so the user has the control is a good idea in general.

We discussed this for a bit an thought of a different approach here. Instead giving the control over exceptions here we could enforce the user to use a `ConstraintViolationException` in the transformation.

In general we are using `ConstraintViolationException` in the source code of the Refinery to signal that something gone wrong in the transformation process.

What about to enforce the user to use the `ConstraintViolationException` in his custom transformation:

```php
trait DeriveTransformFromApplyTo
{
	/**
	 * @param mixed $from
	 * @return Result
	 */
	public function transform($from)
	{
		/** @var Result $result */
		$result = $this->applyTo(new Result\Ok($from));
		if (true === $result->isError()) {
			$error = $result->error();

			if ($error instanceof ConstraintViolationException) {
				throw $error;
			}

			if ($error instanceof \Exception) {
				throw new NotDefinedErrorException();  // Name not final
			}

			throw new ConstraintViolationException($error, 'error');
		}
		return $result->value();
	}
}
```
This would enforce the user to throw always a `ConstraintViolationException` instead of allowing her/him to throw other exceptions.

These were just our thoughts, but we are also fine with your solution to just throw it if its an exception :+1: